src/gui/sdlmain.cpp

```diff
@@ -340,26 +219,156 @@ static GLuint build_shader_gl(GLenum type, const std::string& source)
 	}
 }
 
-static bool load_shader_gl(const std::string& source, GLuint& vertex_out,
-                           GLuint& fragment_out)
+// Build a OpenGL shader program.
+//
+// Input GLSL source must contain both vertex and fragment stages inside their
+// respective preprocessor definitions.
+//
+// Returns a ready to use OpenGL shader program, or zero on failure.
+//
+static GLuint build_shader_program(const std::string& source)
 {
 	if (source.empty()) {
-		return false;
+		LOG_ERR("OPENGL: No shader source present");
+		return 0;
+	}
+
+	auto vertex_shader = build_shader_gl(GL_VERTEX_SHADER, source);
+	if (!vertex_shader) {
+		LOG_ERR("OPENGL: Failed compiling vertex shader");
+		return 0;
 	}
 
-	GLuint s = build_shader_gl(GL_VERTEX_SHADER, source);
-	if (s) {
-		vertex_out = s;
+	auto fragment_shader = build_shader_gl(GL_FRAGMENT_SHADER, source);
+	if (!fragment_shader) {
+		LOG_ERR("OPENGL: Failed compiling fragment shader");
+		glDeleteShader(vertex_shader);
+		return 0;
+	}
+
+	const GLuint shader_program = glCreateProgram();
+
+	if (!shader_program) {
+		LOG_ERR("OPENGL: Failed creating shader program");
+		glDeleteShader(vertex_shader);
+		glDeleteShader(fragment_shader);
+		return 0;
+	}
 
-		s = build_shader_gl(GL_FRAGMENT_SHADER, source);
+	glAttachShader(shader_program, vertex_shader);
+	glAttachShader(shader_program, fragment_shader);
 
-		if (s) {
-			fragment_out = s;
-			return true;
+	glLinkProgram(shader_program);
+
+	glDeleteShader(vertex_shader);
+	glDeleteShader(fragment_shader);
+
+	// Check the link status
+	GLint is_program_linked = GL_FALSE;
+	glGetProgramiv(shader_program, GL_LINK_STATUS, &is_program_linked);
+
+	// The info log might contain warnings and info messages even if the
+	// linking was successful, so we'll always log it if it's non-empty.
+	GLint log_length_bytes = 0;
+	glGetProgramiv(shader_program, GL_INFO_LOG_LENGTH, &log_length_bytes);
+
+	if (log_length_bytes > 1) {
+		std::vector<GLchar> info_log(static_cast<size_t>(log_length_bytes));
+
+		glGetProgramInfoLog(shader_program,
+		                    log_length_bytes,
+		                    nullptr,
+		                    info_log.data());
+
+		if (is_program_linked) {
+			LOG_WARNING("OPENGL: Program info log:\n %s",
+			            info_log.data());
+		} else {
+			LOG_ERR("OPENGL: Failed linking shader program:\n %s",
+			        info_log.data());
 		}
-		glDeleteShader(vertex_out);
 	}
-	return false;
+
+	if (!is_program_linked) {
+		glDeleteProgram(shader_program);
+		return 0;
+	}
+
```

Comment:
Traced the new warning I'm getting `OPENGL: Failed getting double buffering status: OpenGL error: 00000502` back to here. First bad commit is 1d751f521216c9130ce2c8bd46f8f401095e5526

Fix is after this block (we know the program is linked) add `glUseProgram(shader_program);`. This is the way it was before refactoring this into a separate function. After the refactor, you wait to `glUseProgram()` until the function returns but I believe several of the `gl*` functions following this depend on the new program being in use.

Reply:
Cheers, that's good feedback. Yeah, I'm trying to slowly untangle all this mess and not do too much in any single commit. Will fix that today.